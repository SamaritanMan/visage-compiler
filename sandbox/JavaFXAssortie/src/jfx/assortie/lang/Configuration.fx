package jfx.assortie.lang;

import jfx.assortie.system.Module;
import jfx.assortie.system.Sample;

Module {
    name: "JavaFX"
    modules: [
    "jfx.assortie.lang.api.Configuration",
    ]
    samples: [
        Sample{
            name: "New"
            className: "jfx.assortie.samples.empty.Empty"
            visible: true
        }
    ]

}
