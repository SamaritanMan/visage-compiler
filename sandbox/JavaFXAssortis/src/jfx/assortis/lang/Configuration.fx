package jfx.assortis.lang;

import jfx.assortis.system.Module;
import jfx.assortis.system.Sample;

Module {
    name: "JavaFX"
    modules: [
    "jfx.assortis.lang.api.Configuration",
    "jfx.assortis.lang.animation.Configuration",
    ]
    samples: [
        Sample{
            name: "New"
            className: "jfx.assortis.samples.empty.Empty"
            visible: true
        }
    ]

}
