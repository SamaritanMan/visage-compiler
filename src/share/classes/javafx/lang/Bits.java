package javafx.lang;

/**
 *  Bit operations
 */

public class Bits {

    public static int shiftLeft(int value, int shift) {
        return value << shift;
    }

    public static int shiftRight(int value, int shift) {
        return value >> shift;
    }

    public static int bitOr(int value, int mask) {
        return value | mask;
    }

    public static int bitAnd(int value, int mask) {
        return value & mask;
    }
    
    public static int bitXor(int value, int mask) {
        return value ^ mask;
    }

    public static int complement(int value) {
        return ~value;
    }

    public static int add(int value, int mask) {
        return value | mask;
    }

    public static int remove(int value, int mask) {
        return value & ~mask;
    }

    public static boolean contains(int value, int mask) {
        return (value & mask) != 0;
    }


    public static long shiftLeft(long value, int shift) {
        return value << shift;
    }

    public static long shiftRight(long value, int shift) {
        return value >> shift;
    }

    public static long bitOr(long value, long mask) {
        return value | mask;
    }

    public static long bitAnd(long value, long mask) {
        return value & mask;
    }
    
    public static long bitXor(long value, long mask) {
        return value ^ mask;
    }

    public static long complement(long value) {
        return ~value ;
    }

    public static long add(long value, long mask) {
        return value | mask;
    }

    public static long remove(long value, long mask) {
        return value & ~mask;
    }

    public static boolean contains(long value, long mask) {
        return (value & mask) != 0;
    }

}
