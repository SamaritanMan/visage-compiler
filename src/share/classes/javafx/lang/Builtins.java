/*
 * Copyright 2008-2009 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package javafx.lang;

import com.sun.javafx.runtime.FXBase;

/**
 * These functions are automaticlly imported for
 * all JavaFX Scripts to use
 *
 * @author Brian Goetz
 * @author Saul Wold
 * @profile common
 */
public class Builtins {
    /**
     * Compare 2 JavaFX Script Objects
     * 
     * @param a The first object to be compared
     * @param b the second object to be compared
     * @return true if they are the same Object
     */
    public static boolean isSameObject(Object a, Object b) {
        return a == b;
    }

    /**
     * Print the Object 'val'.
     * 
     * @param val The Object to be printed
     */
    public static void print(Object val) {
        if (val == null) {
            System.out.print(val);
        } else if (val instanceof String) {
            System.out.print((String) val);
        } else {
            System.out.print(val.toString());
        }
    }

    /**
     * Print the Object 'val' and a new-line.
     * 
     * @param val The Object to be printed
     */
    public static void println(Object val) {
        if (val == null) {
            System.out.println(val);
        } else if (val instanceof String) {
            System.out.println((String) val);
        } else {
            System.out.println(val.toString());
        }
    }

    /**
     * Test if an instance variable has been initialized.
     * 
     * @param args The variable to be tested (args[0] is the FXObject instance,
     * args[1] is the var offset).
     * @return true if the object has been initialized already
     */
    public static boolean isInitialized(Object... args) {
        //should return false until flag is correctly set - otherwise
        //this will always return true - which causes some reg failures
        //the code below should be uncommented when the flag is set properly
        return false;
        /*
        FXBase inst = (FXBase)args[0];
        int offset = (Integer)args[1];
        return inst.varTestBits$(offset,
                FXBase.VFLGS$IS_INITIALIZED,
                FXBase.VFLGS$IS_INITIALIZED);
         */
    }
}
