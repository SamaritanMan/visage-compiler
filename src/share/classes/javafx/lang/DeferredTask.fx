/*
 * Copyright 2008 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package javafx.lang;
import com.sun.javafx.runtime.Entry;

/**
  * A {@code DeferredTask} represents an action that should be executed at a
  * later time of the system's choosing.
  * <p />
  * In systems based on event dispatch, such as Swing, execution of a
  * {@code DeferredTask} generally means putting it on the event queue
  * for later processing.
  *
  * @profile common
  */
public class DeferredTask {
    /** 
     * A function to be called when the deferred task is executed.
     * <p />
     * Must be provided.
     *
     * @profile common
     */
    public /* required initonly */ var action: function() : Void;

    /** 
     * A {@code boolean} value indicating whether the deferred task has actually
     * been run, which can be bound to.  Read-only. 
     *
     * @profile common
     * @readonly
     */
    public /* readable private */ var done : Boolean;

    postinit {
        Entry.deferTask(function() : Void { try { action(); } finally { done = true } });
    }
}
