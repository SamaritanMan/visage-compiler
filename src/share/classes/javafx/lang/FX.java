/*
 * Copyright 2008 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package javafx.lang;

import java.util.Vector;
import com.sun.javafx.runtime.Entry;
import com.sun.javafx.functions.Function0;
import com.sun.javafx.runtime.SystemProperties;
import com.sun.javafx.runtime.FXExit;

import com.sun.javafx.runtime.sequence.Sequence;

/**
 * FX, analogous to java.lang.System, is a place to store static utility methods.  
 *
 * @author Brian Goetz
 * @author Saul Wold
 * 
 * @profile common
 */
public class FX {

    public static boolean isSameObject(Object a, Object b) {
        return Builtins.isSameObject(a, b);
    }

    /**
     * Print the Object 'val'.
     * 
     * @param val The Object to be printed
     */
    public static void print(Object val) {
        Builtins.print(val);
    }

    /**
     * Print the Object 'val' and a new-line.
     * 
     * @param val The Object to be printed
     */
    public static void println(Object val) {
        Builtins.println(val);
    }

    /**
     * Test if an instance variable has been initialized.
     * 
     * @param varRef The variable to be tested.
     */
    public static boolean isInitialized(Object varRef) {
        return Builtins.isInitialized(varRef);
    }

    /**
     * Retrieve System Property.
     * System Properties in JavaFX environment can be classified into 3 types:
     * 1. Runtime platform associated property. 
     *    Those properties have an equivalent in current java runtime 
     *    environment (SE/ME). The FX.getProperty() method retrieves
     *    those properties by mapping specified key with runtime platform key. 
     * </p>
     * 2. JavaFX specific property. 
     *    Those properties are specific to JavaFX environment therefore
     *    value of the properties is specified in the JavaFX tree.
     * </p>
     * 3. The property has no association with current runtime platform 
     *    nor is JavaFX specific, therefore those properties are not 
     *    supported and null is returned.
     * @param key Environment Property to be inquired
     * @return The string value of the property
     * @throws SecurityException if a security manager exists and its checkPropertyAccess method doesn't allow access to the specified system property. 
     * @throws NullPointerException if key is null. 
     * @profile common
     */
    public static String getProperty (String key) {
        return SystemProperties.getProperty(key);
    }
 
    /*
     * This static will be unique to each applet, when we move
     * to the FXME/Embedded this will need to be tied the AMS
     * TODO: for Mobile Guys
     */
    private static FXSystemActionData exitData = new FXSystemActionData();

    /**
     * Exits the Script and causes any Shutdown Actions to be called
     * This may cause the Runtime to exit as {@code System.exit()} 
     * depending on the underlying implementation.
     * </p><p> 
     * Any Shutdown Actions that were previously added using the 
     * {@code addShutdownAction()} function will be exectued at this time
     * in LIFO ordering.
     * </p><p>
     * A second call to {@code FX.exit()} once {@code FX.exit()} has 
     * started will result a {@code IllegalStateException} to be thrown,
     * this can occur if a {@code Timeline} calls {@code FX.exit()} while
     * FX.exit is started.
     * If a call to {@code FX.exit()} occurs in a Shutdown Action, that
     * action's function will simply exit without completing the rest of
     * its operation and the next Shutdown Action, if any, will run.
     * </p><p>
     * This function will not normally return to the calling Script.
     * </p>
     *
     * @profile common
     */
    public static void exit() {
        if (exitData.called) {
            throw new IllegalStateException("Can not call FX.exit() twice");
        } else {
            exitData.called = true;
        }
        /*
         * Run the exit actions
         */
        exitData.runActions();

        /*
         * Call to Entry.java in order to get the RuntimeProvider
         * to do additional cleanup as needed for that provider
         */
        Entry.exit();

        /*
         * Mark this as false for handling restarting from the
         * same VM or Browser Context
         */
        exitData.called = false;
        
        /*
         * Use of FXExit here is needed because the EDT
         * will pass it along rather than catch and quit
         */
        throw new FXExit();
    }

    /**
     * Adds an action to the queue to be executed at {@code FX.exit()} time
     * This action will be added to the queue as a push stack, meaning that
     * they will be excuted in FILO ordering. Duplicate actions are
     * not allowed and will cause the orignal Handle to be returned with
     * no reordering.
     * 
     * @param  action of type {@code function():Void} that will be executed
     * at {@code FX.exit()} time. Only one copy of an action can be in 
     * the queue, an attempt to add the same action a second time will
     * return the previous Handle without any reodering.
     * @return Handle used to remove the action if needed.
     *
     * @profile common
     */
    public static int addShutdownAction(Function0<Void> action) {
        if (action == null) {
            throw new NullPointerException("Action function can not be null");
        } else {
            return exitData.addAction(action);
        }
    }

    /**
     * Removes the action from the queue specified by the actionType parameter. 
     *
     * @param  action of type {@code function():Void} that will be removed
     * from the Shutdown Action Stack
     * @return a Boolean value signifing sucess or failure of removing
     * the action
     *
     * @profile common
     */
    public static boolean removeShutdownAction(int handle) {
        return exitData.removeAction(handle);
    }

    /**
     * A {@code deferAction} represents an action that should be executed at a
     * later time of the system's choosing.
     * <p />
     * In systems based on event dispatch, such as Swing, execution of a
     * {@code deferAction} generally means putting it on the event queue
     * for later processing.
     *
     * @param  action of type {@code function():Void} that will be executed
     * later based on the implementation.
     * 
     * @profile common
     */
    public static void deferAction(Function0<Void> action) {
        if (action == null) {
            throw new NullPointerException("Action function can not be null");
        } else {
            Entry.deferAction(action);
        }
    }

    /**
     * For JavaFX Script applications that are started on the command
     * line,running application. Returns null if there were no incoming
     * arguments or if this application was not invoked from the
     * command line. 
     *
     * @profile common
     */
    public static Sequence<String> getArguments() {
        return Entry.getArguments();
    }

    /**
     * Returns the named incoming argument for the current JavaFX
     * Script program; this is used for certain environments (in
     * particular, applets) where incoming arguments are represented
     * as name/value pairs. This usually returns a String, but some
     * environments may return other kinds of values. Accepts numbers
     * in the form of Strings (e.g. {@code getArgument("0")}) to
     * provide unification with {@link #getArguments getArguments}.
     * Returns null if the given named argument does not exist.
     *
     * @profile common
     */
    public static Object getArgument(String key) {
        return Entry.getArgument(key);
    }

    /*
     * This inner help class is used to store the Action Data needed 
     * for exitActions, in the future there may me addition System 
     * Actions that will be required such as Low Resources or ...
     */
    private static class FXSystemActionData {
        boolean called;
        Vector actions, handles;

        private static FXSystemActionData singleton = null;

        private FXSystemActionData() {
            called = false;
            actions = new Vector();
            handles = new Vector();
        }

        int addAction(Object action) {
            int hash = action.hashCode();

            // Check for action already in Vector
            int index = handles.indexOf(hash);
            if (index != -1) {
                // Return the hash without reorder
                return hash;
            }
            actions.addElement(action);
            handles.addElement(hash);
            return hash;
        }

        boolean removeAction(int handle) {
            if (handles.isEmpty() || !handles.contains(handle)) {
                return false;
            } else {
                try {
                    int index = handles.indexOf(handle);
                    if (index != -1) {
                        actions.removeElementAt(index);
                        handles.removeElementAt(index);
                    }
                } catch (ArrayIndexOutOfBoundsException e) {
                    return false;
                }
                return true;
            }
        }

        void runActions() {
            if (actions == null) {
                return;
            }
            while (!actions.isEmpty()) {
                Function0<Void> action = null;
                // Execute LIFO Action
                action = (Function0<Void>) actions.lastElement();
                actions.remove(actions.lastIndexOf(action));
                if (action != null) {
                    try {
                        /*
                         * TODO: add timer to kill long running Action
                         */
                        action.invoke();
                    } catch (Throwable ignore) {
                        // Ignore all Throwables
                    }
                }
            }
            handles.removeAllElements();
        }
    }
}
