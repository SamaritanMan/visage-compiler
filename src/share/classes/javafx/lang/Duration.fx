/*
 * Copyright 1999-2007 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Sun designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Sun in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package javafx.lang;

import java.lang.Object;
import java.lang.Comparable;
import java.lang.Long;
import java.lang.Math;

/**
 * A class that defines a duration of time.  Duration instances are defined in
 * milliseconds, but can be easily created using time literals; for
 * example, a two-and-a-half minute Duration instance can be defined in several
 * ways:
 * <code><pre>    Duration t = 2m + 30s;
    Duration t = 2.5s;
    Duration t = 2500ms;</pre></code>
 */
public class Duration extends Comparable {

    /** The duration of time, as expressed in milliseconds. */
    public attribute millis: Number;

    public function equals(obj:Object):Boolean {
        if (obj instanceof Duration) {
            var t = obj as Duration;
            return t.millis == millis;
        }
        return false;
    }

    public function compareTo(obj:Object):Integer {
        var t = obj as Duration;
        var m1 = millis;
        var m2 = t.millis;
        var cmp = m1 - m2;
        return if (cmp < 0) -1 else if (cmp > 0) 1 else 0;
    }

    public function hashCode():Integer {
        return Long.valueOf(millis.longValue()).hashCode();
    }

    /** Returns the number of milliseconds in this period. */
    public function toMillis():Number {
        return millis;
    }
    
    /** Returns the number of whole seconds in this period. */
    public function toSeconds():Number {
        return Math.floor(millis / 1000);
    }
    
    /** Returns the number of whole minutes in this period. */
    public function toMinutes(): Number {
        return Math.floor(millis / 60 / 1000);
    }
    
    /** Returns the number of whole hours in this period. */
    public function toHours(): Number {
        return Math.floor(millis / 60 / 60 / 1000);
    }

    /** Add this instance and another Duration instance to return a new Duration instance.
     *  This function does not change the value of called Duration instance. */
    public function add(other:Duration):Duration {
        return Duration {
            millis: millis + other.millis;
        }
    }

    /** Subtract this instance from another Duration instance to return a new Duration instance.
     *  This function does not change the value of called Duration instance. */
    public function sub(other:Duration):Duration {
        return Duration {
            millis: millis - other.millis;
        }
    }

    /** Multiply this instance with a number to return a new Duration instance.
     *  This function does not change the value of called Duration instance. */
    public function mul(n:Number):Duration {
        return Duration {
            millis: millis * n;
        }
    }


    /** Divide this instance by a number to return a new Duration instance.
     *  This function does not change the value of called Duration instance. */
    public function div(n:Number):Duration {
        return Duration {
            millis: millis / n;
        }
    }

    public function negate():Duration {
        return Duration {
            millis: -millis;
        }
    }

    public function toString(): String {
        return "{millis}ms";
    }

    public function lt(other: Duration):Boolean {
        return compareTo(other) < 0;
    }

    public function le(other: Duration):Boolean {
        return compareTo(other) <= 0;
    }

    public function gt(other: Duration):Boolean {
        return compareTo(other) > 0;
    }

    public function ge(other: Duration):Boolean {
        return compareTo(other) >= 0;
    }

    public function toDate():java.util.Date {
        return new java.util.Date(millis.longValue());
    }
}
