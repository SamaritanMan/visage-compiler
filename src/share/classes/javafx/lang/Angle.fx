/*
 * Copyright 2008-2009 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package javafx.lang;

import java.lang.Comparable;
import java.lang.IllegalStateException;
import java.lang.Math;
import java.lang.ThreadLocal;
import java.text.DecimalFormat;
import java.util.Formattable;
import javafx.util.Bits;
import java.util.Formatter;
import java.util.FormattableFlags;
import java.lang.StringBuilder;

/**
 * Used to specify a angle of 0ms.
 *
 * @profile common
 */
public def ZERO:Angle = Angle {value: 0, unit: AngleUnit.RADIAN}

def DF = ThreadLocal {
    override function initialValue():Object {
        return new DecimalFormat("#.######");
    }
}

// script-level "static" functions below

/**
 * Factory method that returns a Angle instance for a specified
 * value and unit.
 *
 * @param value the magnitude of the angle
 * @param unit the unit of the angle
 * @return a Angle instance of the specified value and unit type
 * @profile common
 */
public function valueOf(value:Double, unit:AngleUnit):Angle {
    if (value == 0)
        ZERO
    else Angle {
        value: value
        unit: unit
    }
}

public function valueOf(value:Float, unit:AngleUnit):Angle {
    valueOf(value as Double, unit)
}

public def TYPE_INFO = com.sun.javafx.runtime.TypeInfo.makeAndRegisterTypeInfo(ZERO);

/**
 * A class that defines an angle measurement.  Angle instances can be specified
 * as degrees, radians, or turns.  For precise calculations, they are stored
 * in the given unit type as a double value.
 * <p>
 * Angles can be easily created using angle literals; for
 * example, a 90 degree Angle instance can be defined in several
 * ways:
 * <code><pre>    Angle a = 90deg;
    Angle a = .25turn;
    Angle a = Angle.valueOf(Math.PI / 2, AngleUnit.RADIAN);</pre></code>
 * Angle instances are immutable, and are therefore replaced rather than modified.
 * To create a new Angle instance, either use a time literal, or use the
 * <code>Angle.valueOf(value, unit)</code> factory method.
 *
 * @profile common
 */
public class Angle extends Comparable, Formattable {

    var value:Double;
    var unit:AngleUnit;

    /**
     * Returns the number of degrees in this angle.
     *
     * @profile common
     */
    public function toDegrees():Double {
        if (unit == AngleUnit.DEGREE) {
            return value;
        } else if (unit == AngleUnit.RADIAN) {
            return value * 180 / Math.PI;
        } else if (unit == AngleUnit.TURN) {
            return value * 360;
        } else {
            throw new IllegalStateException("Unknown unit: {unit}");
        }
    }

    /**
     * Returns a new Length object with an angle in degrees.
     *
     * @profile common
     */
    public function toDegreeAngle():Angle {
        if (this == ZERO) this;
        Angle {
            value: toDegrees();
            unit: AngleUnit.DEGREE;
        }
    }
    
    /**
     * Returns the number of radians in this angle.
     *
     * @profile common
     */
    public function toRadians():Number {
        if (unit == AngleUnit.RADIAN) {
            return value;
        } else if (unit == AngleUnit.DEGREE) {
            return value * Math.PI / 180;
        } else if (unit == AngleUnit.TURN) {
            return value * 2 * Math.PI;
        } else {
            throw new IllegalStateException("Unknown unit: {unit}");
        }
    }
    
    /**
     * Returns a new Length object with an angle in radians.
     *
     * @profile common
     */
    public function toRadianAngle():Angle {
        if (this == ZERO) this;
        Angle {
            value: toRadians();
            unit: AngleUnit.RADIAN;
        }
    }

    /**
     * Returns the number of turns in this angle.
     *
     * @profile common
     */
    public function toTurns():Number {
        if (unit == AngleUnit.TURN) {
            return value;
        } else if (unit == AngleUnit.DEGREE) {
            return value / 360;
        } else if (unit == AngleUnit.RADIAN) {
            return value / 2 / Math.PI;
        } else {
            throw new IllegalStateException("Unknown unit: {unit}");
        }
    }

    /**
     * Returns a new Length object with an angle in turns.
     *
     * @profile common
     */
    public function toTurnAngle():Angle {
        if (this == ZERO) this;
        Angle {
            value: toTurns();
            unit: AngleUnit.TURN;
        }
    }

    /**
     * Add this instance and another Angle instance to return a new Angle instance.
     * This function does not change the value of the called Angle instance.
     *
     * @profile common
     */
    public function add(other:Angle):Angle {
        if (unit == other.unit) {
            return valueOf(value + other.value, unit);
        } else {
            return valueOf(toRadians() + other.toRadians(), AngleUnit.RADIAN);
        }
    }

    /**
     * Subtract other Angle instance from this instance to return a new Angle instance.
     * This function does not change the value of the called Angle instance.
     *
     * @profile common
     */
    public function sub(other:Angle):Angle {
        if (unit == other.unit) {
            return valueOf(value - other.value, unit);
        } else {
            return valueOf(toRadians() - other.toRadians(), AngleUnit.RADIAN);
        }
    }

    /**
     * Multiply this instance with a number to return a new Angle instance.
     * This function does not change the value of the called Angle instance.
     *
     * @profile common
     */ 
    public function mul(n:Number):Angle {
        return valueOf(value * n, unit);
    }


    /**
     * Divide this instance by a number to return a new Angle instance.
     * This function does not change the value of the called Angle instance.
     *
     * @profile common
     */     
    public function div(n:Number):Angle {
        if (n == 0) {
            throw new java.lang.ArithmeticException("/ by zero");
        }
        
        return valueOf(value / n, unit);
    }

    /**
     * Divide this instance by another Angle to return the ratio.
     * This function does not change the value of the called Angle instance.
     *
     * @profile common
     */     
    public function div(other:Angle):Double {
        if (other.value == 0) {
            throw new java.lang.ArithmeticException("/ by zero");
        }
        if (unit == other.unit) {
            return value / other.value;
        } else {
            return toRadians() / other.toRadians();
        }
    }

    /** 
     * Return a new Angle instance which has a negative number of milliseconds
     * from this instance.  For example, <code>(50deg).negate()</code> returns
     * a Angle of -50 degrees.
     * This function does not change the value of the called Angle instance.
     *
     * @profile common
     */
    public function negate():Angle {
        return valueOf(-value, unit);
    }

    /**
     * Prints out a string representation of the length with a precision of
     * up to 6 digits (internally angles are represented as doubles with much
     * higher precision).
     *
     * @profile common
     */        
    override function toString():String {
        def df = DF.get() as DecimalFormat;
        df.setMinimumFractionDigits(0);
        return "{df.format(value)}{unit.getSuffix()}";
    }

    /**
     * Formatting function that allows angles to be properly formatted by
     * printf, String.format, or the JavaFX format primitive.
     * <p>
     * Width sets the minimum width of this field, following the justification
     * rules set in flags.
     * <p>
     * Precision can be used to specify exactly how many digits should appear
     * to the right of the decimal, including zero padding.
     * <p>
     * The alternate form is not supported for angles.
     * <p>
     * Finally, the uppercase flag is also supported, but only applies to the
     * length suffix.
     */
    override function formatTo(formatter:Formatter, flags:Integer, width:Integer, precision:Integer):Void {
        def uppercase = Bits.bitAnd(flags, FormattableFlags.UPPERCASE) != 0;
        def leftJustify = Bits.bitAnd(flags, FormattableFlags.LEFT_JUSTIFY) != 0;
        def sb = StringBuilder {};
        def df = DF.get() as DecimalFormat;
        df.setMinimumFractionDigits(precision);
        sb.append(df.format(value));
        sb.append(if (uppercase) unit.getSuffix().toUpperCase() else unit.getSuffix());
        def len = sb.length();
        if (len < width) {
            if (leftJustify) {
                for (i in [1..width-len]) sb.append(' ');
            } else {
                for (i in [1..width-len]) sb.<<insert>>(0, ' ');
            }
        }
        formatter.out().append(sb);
    }

    /**
     * Returns true if the specified angle is less than (<) this instance.
     * INDEFINITE is treated as if it were positive infinity.
     *
     * @profile common
     */
    public function lt(other:Angle):Boolean {
        return compareTo(other) < 0;
    }

    /** 
     * Returns true if the specified angle is less than or equal to (<=) this instance.
     * INDEFINITE is treated as if it were positive infinity.
     *
     * @profile common
     */
    public function le(other:Angle):Boolean {
        return compareTo(other) <= 0;
    }

    /** 
     * Returns true if the specified angle is greater than (>) this instance.
     * INDEFINITE is treated as if it were positive infinity.
     *
     * @profile common
     */
    public function gt(other:Angle):Boolean {
        return compareTo(other) > 0;
    }

    /**
     * Returns true if the specified angle is greater than or equal to (>=) this instance.
     * INDEFINITE is treated as if it were positive infinity.
     *
     * @profile common
     */
    public function ge(other:Angle):Boolean {
        return compareTo(other) >= 0;
    }

    override function compareTo(obj:Object):Integer {
        def a = obj as Angle;
        if (a.value == value and a.unit == unit) {
            return 0;
        }
        def cmp = toRadians() - a.toRadians();
        return if (cmp < 0)
             -1
        else if(cmp > 0)
             1
        else
             0
    }

    override function equals(obj:Object):Boolean {
        if (obj instanceof Angle) {
            def a = obj as Angle;
            if (isSameObject(a, this)) {
                return true;
            }
            return a.toRadians() == toRadians();
        }
        return false;
    }

    override function hashCode():Integer {
        var v = value as Long;
        v = Bits.bitXor(v, unit.ordinal());
        return Bits.bitXor(v, Bits.unsignedShiftRight(v, 32)) as Integer;
    }
}
