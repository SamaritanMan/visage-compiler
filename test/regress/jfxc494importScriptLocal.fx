/**
 * Regression test for JFXC-494 : Script importing
 *
 * @compilefirst jfxc494subloca.fx
 * @test
 * @run
 */

import jfxc494subloca.*;
import java.lang.System;

System.out.println(pkgDef);
System.out.println(pkgFunction());
var x = new pkgClass;
System.out.println(x.thing);

