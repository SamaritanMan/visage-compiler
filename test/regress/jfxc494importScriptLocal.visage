/**
 * Regression test for JFXC-494 : Script importing
 *
 * @compilefirst jfxc494subloca.visage
 * @test
 * @run
 */

import jfxc494subloca.*;
import java.lang.System;

System.out.println(pkgDef);
System.out.println(pkgFunction());
var x = new pkgClass;
System.out.println(x.thing);

