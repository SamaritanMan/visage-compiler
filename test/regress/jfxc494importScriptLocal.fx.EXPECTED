Close I am.
Close here too.
